content/blog-posts/2024-06-13-ebpf-licensing/index.md

```diff
@@ -0,0 +1,50 @@
+
+If an eBPF program needs to be GPL licensed, what does that mean for your userspace executables? Can a non-GPL'd user space executable bundle/load/interact with GPL licensed eBPF bytecode running inside the kernel? Yes! The intent is for this to be normal, reasonable practice. All of the userspace interactions fall under a Linux kernel syscall API exception. All applications running in user space interact with the kernel using the syscall interface, and the same is true for eBPF userspace applications.
+
+With the GPL [derived works](https://copyleft.org/guide/comprehensive-gpl-guide.html), external code that is either [linked](https://www.gnu.org/licenses/old-licenses/gpl-2.0-faq.html#LinkingWithGPL) statically at build time or dynamically at runtime, and modular programs (single applications that may execute separately but depend on detailed communication of internal data structures or shared memory), must be distributed in a GPL compatible manner. However, userspace executables for eBPF programs are none of the above.
+
+The Linux kernel developers have made a very deliberate effort to provide documented intent around the issue of what compromises a derived work of the Linux kernel and what does not.  They have designated the set of system calls as the syscall API that can be called from other programs without those programs being considered a derived work of the kernel.
+
+A userspace executable and the related eBPF bytecode is generally considered an aggregate of the two pieces of software. They are two separate works that interact via the Linux syscall API to get useful work done and can be packaged together without being considered modular parts of a single application derived from the Linux kernel. Aggregates are a number of separate programs, distributed together that communicate with each other as if they were independent executables on the system (using standard OS provided mechanisms such as pipes or file descriptors). The GPLv2 [FAQ](https://www.gnu.org/licenses/old-licenses/gpl-2.0-faq.en.html#TOCMereAggregation) lays out that the creation and that distribution of an aggregate, of GPL and non-GPL compatible code, should be permitted.
+
+![modular vs aggregate code](/modular-vs-aggregate.png)
+
+eBPF programs have no ability to link either statically or dynamically with user space code. User space applications interact with the in-kernel eBPF program exclusively through the syscall interface, in the same way they interact with any other aspect of the kernel, and thus are not derivative work. This is expressly described in the [Linux syscall note](https://github.com/torvalds/linux/blob/master/LICENSES/exceptions/Linux-syscall-note) and further clarified in the eBPF licensing [section](https://docs.kernel.org/bpf/bpf_licensing.html) “Generally, proprietary-licensed applications and GPL licensed BPF programs written for the Linux kernel in the same package can co-exist because they are separate executable processes.”
```

Comment:
```suggestion
eBPF programs have no ability to link either statically or dynamically with user space code. User space applications interact with the in-kernel eBPF program exclusively through the syscall interface, in the same way they interact with any other aspect of the kernel, and thus are not derivative works. This is expressly described in the [Linux syscall note](https://github.com/torvalds/linux/blob/master/LICENSES/exceptions/Linux-syscall-note) and further clarified in the eBPF licensing [section](https://docs.kernel.org/bpf/bpf_licensing.html) “Generally, proprietary-licensed applications and GPL licensed BPF programs written for the Linux kernel in the same package can co-exist because they are separate executable processes.”
```